### PR TITLE
feat: Deterministic mode for filtered topK kernel

### PR DESCRIPTION
**Pull request by Linda-Stadter**
## 📌 Description

Adds `deterministic=True` parameter to `flashinfer.top_k()` for bitwise-reproducible output. Optional and defaults to `False`.

python
``values, indices = flashinfer.top_k(logits, k, deterministic=True)``

When `deterministic=True`:
- Forces the FilteredTopK algorithm (k ≤ 2048, 128KB smem required). Raises  `RuntimeError` if preconditions aren't met.
- Boundary elements that equal the pivot value are collected via a deterministic ballot-scan instead of non-deterministic `atomicAdd`.
- Post-kernel `SortTopKByIndexKVKernel` enforces deterministic order by sorting output indices ascending.
- For `sorted=True`, an additional stable value sort produces (value desc, index asc) ordering. The sort through ``torch.sort`` in ``topk.py`` is then disabled
- Further fixes uint32 overflow issue on large batches when ``batch_size * vocab_size > 2^32``

## Benchmark

B200
```Algorithm: filtered
  batch_size=4096  vocab_size=200000  dtype=fp32

       |                --- Unsorted ---                 |                 --- Sorted ---
     k |     NonDet    GB/s        Det    GB/s  overhead  |     NonDet    GB/s        Det    GB/s  overhead
-----------------------------------------------------------------------------------------------------------
   128 |    1.838ms  1783.0    1.894ms  1730.3     +3.0%  |    1.885ms  1738.3    1.912ms  1713.6     +1.4%
   256 |    1.846ms  1774.9    1.913ms  1712.9     +3.6%  |    1.956ms  1674.8    1.940ms  1689.3     -0.9%
   512 |    1.974ms  1659.6    2.054ms  1595.7     +4.0%  |    2.096ms  1563.6    2.103ms  1558.4     +0.3%
  1024 |    1.994ms  1643.5    2.105ms  1556.8     +5.6%  |    2.145ms  1527.3    2.202ms  1488.0     +2.6%
  2048 |    2.379ms  1377.4    2.581ms  1269.6     +8.5%  |    2.658ms  1232.8    2.782ms  1178.0     +4.7%

Algorithm: filtered
  batch_size=8192  vocab_size=200000  dtype=fp32

       |                --- Unsorted ---                 |                 --- Sorted ---
     k |     NonDet    GB/s        Det    GB/s  overhead  |     NonDet    GB/s        Det    GB/s  overhead
-----------------------------------------------------------------------------------------------------------
   128 |    3.652ms  1794.8    3.765ms  1740.6     +3.1%  |    3.723ms  1760.2    3.797ms  1725.9     +2.0%
   256 |    3.622ms  1809.2    3.749ms  1748.0     +3.5%  |    3.813ms  1718.7    3.800ms  1724.4     -0.3%
   512 |    3.939ms  1663.6    4.088ms  1603.2     +3.8%  |    4.155ms  1577.1    4.181ms  1567.4     +0.6%
  1024 |    3.973ms  1649.7    4.195ms  1562.1     +5.6%  |    4.265ms  1536.5    4.382ms  1495.6     +2.7%
  2048 |    4.747ms  1380.5    5.159ms  1270.2     +8.7%  |    5.272ms  1243.2    5.556ms  1179.5     +5.4%

Algorithm: filtered
  batch_size=16384  vocab_size=200000  dtype=fp32

       |                --- Unsorted ---                 |                 --- Sorted ---
     k |     NonDet    GB/s        Det    GB/s  overhead  |     NonDet    GB/s        Det    GB/s  overhead
-----------------------------------------------------------------------------------------------------------
   128 |    7.250ms  1808.0    7.462ms  1756.6     +2.9%  |    7.366ms  1779.4    7.526ms  1741.7     +2.2%
   256 |    7.189ms  1823.3    7.430ms  1764.0     +3.4%  |    7.545ms  1737.2    7.524ms  1742.0     -0.3%
   512 |    7.806ms  1679.1    8.101ms  1617.9     +3.8%  |    8.232ms  1592.3    8.279ms  1583.1     +0.6%
  1024 |    7.879ms  1663.7    8.332ms  1573.1     +5.8%  |    8.432ms  1554.5    8.695ms  1507.4     +3.1%
  2048 |    9.416ms  1392.0   10.180ms  1287.5     +8.1%  |   10.421ms  1257.8   10.965ms  1195.4     +5.2%
```

<details>
<summary>Benchmark code</summary>

```import os

import numpy as np
import torch

import flashinfer
from flashinfer.testing.utils import bench_gpu_time

BATCH_SIZES = [4096, 8192, 16384]
VOCAB_SIZE = 200000
K_VALUES = [128, 256, 512, 1024, 2048]
ELEM_SIZE = 4  # fp32

ALGOS = ["filtered"]


def bench_one(fn):
    times = bench_gpu_time(fn, enable_cupti=True, dry_run_iters=10, repeat_iters=100)
    return float(np.median(times))


def input_gbps(time_ms, batch_size):
    return (batch_size * VOCAB_SIZE * ELEM_SIZE) / (time_ms * 1e-3) / 1e9


def fmt_overhead(test_ms, base_ms):
    if base_ms == 0:
        return "    N/A"
    pct = (test_ms / base_ms - 1) * 100
    return f"{pct:>+7.1f}%"


def print_row(
    k, nd_ms, nd_gbps, d_ms, d_gbps, oh, nd_s_ms, nd_s_gbps, d_s_ms, d_s_gbps, oh_s
):
    print(
        f"  {k:>4} |"
        f"  {nd_ms:>7.3f}ms {nd_gbps:>7.1f}"
        f"  {d_ms:>7.3f}ms {d_gbps:>7.1f}"
        f"  {oh:>8}"
        f"  |"
        f"  {nd_s_ms:>7.3f}ms {nd_s_gbps:>7.1f}"
        f"  {d_s_ms:>7.3f}ms {d_s_gbps:>7.1f}"
        f"  {oh_s:>8}"
    )


@torch.inference_mode()
def main():
    dtype = torch.float32

    for algo in ALGOS:
        if algo == "auto":
            os.environ.pop("FLASHINFER_TOPK_ALGO", None)
        else:
            os.environ["FLASHINFER_TOPK_ALGO"] = algo

        for batch_size in BATCH_SIZES:
            scores = torch.randn(batch_size, VOCAB_SIZE, device="cuda", dtype=dtype)

            print(f"Algorithm: {algo}")
            print(f"  batch_size={batch_size}  vocab_size={VOCAB_SIZE}  dtype=fp32")
            print()

            half = 48
            print(f"       |{'--- Unsorted ---':^{half}} |{'--- Sorted ---':^{half}}")
            print(
                f"     k |"
                f"  {'NonDet':>9} {'GB/s':>7}"
                f"  {'Det':>9} {'GB/s':>7}"
                f"  {'overhead':>8}"
                f"  |"
                f"  {'NonDet':>9} {'GB/s':>7}"
                f"  {'Det':>9} {'GB/s':>7}"
                f"  {'overhead':>8}"
            )
            print("-" * (7 + 1 + half + 2 + 1 + half))

            for k in K_VALUES:
                nd_unsorted = bench_one(
                    lambda k=k, s=scores: flashinfer.top_k(s, k, sorted=False)
                )
                nd_sorted = bench_one(
                    lambda k=k, s=scores: flashinfer.top_k(s, k, sorted=True)
                )

                d_unsorted = bench_one(
                    lambda k=k, s=scores: flashinfer.top_k(
                        s, k, sorted=False, deterministic=True
                    )
                )
                d_sorted = bench_one(
                    lambda k=k, s=scores: flashinfer.top_k(
                        s, k, sorted=True, deterministic=True
                    )
                )

                print_row(
                    k,
                    nd_unsorted,
                    input_gbps(nd_unsorted, batch_size),
                    d_unsorted,
                    input_gbps(d_unsorted, batch_size),
                    fmt_overhead(d_unsorted, nd_unsorted),
                    nd_sorted,
                    input_gbps(nd_sorted, batch_size),
                    d_sorted,
                    input_gbps(d_sorted, batch_size),
                    fmt_overhead(d_sorted, nd_sorted),
                )

            del scores
            print()

    os.environ.pop("FLASHINFER_TOPK_ALGO", None)


if __name__ == "__main__":
    main()

```
</details>

## Limitations

- Only supported for FilteredTopK (k ≤ 2048). Not available for the multi-CTA radix path.
- Forces FilteredTopK even when the heuristic would prefer multi-CTA, which may be slower for some configurations
- Not optimized for bf16

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Release Notes

* **New Features**
  * Added deterministic mode for reproducible, bit-identical top-k results across runs.
  * Added sorted output option to control ordering of top-k selection results.

* **Tests**
  * Extensive test coverage added for deterministic reproducibility and sorted output functionality.
  * Tests validating edge cases, tie-breaking behavior, and cross-algorithm correctness.

### CODE REVIEW

**Comment by yzh119** [comment]
Hi @Linda-Stadter is it ready for review?

btw, there is another PR from the community which seems to be working on the same thing: https://github.com/flashinfer-ai/flashinfer/pull/2661, I haven't checked carefully but would you mind also taking a look?

**Comment by Linda-Stadter** [comment]
This PR is now part of https://github.com/flashinfer-ai/flashinfer/pull/2661
